examples/pytorch/tgn.py

```diff
@@ -0,0 +1,252 @@
+    TimeEncoder,
+)
+from torch.utils.data import DataLoader
+
+from deepgnn.graph_engine.data.mooc import MOOC
+from deepgnn import get_logger
+
+device = torch.device("cpu")
+train_batch_size = 200
+num_neighbors = 10
+torch.manual_seed(42)
+
+
+class GraphAttentionEmbedding(torch.nn.Module):
+    """Temporal attention layer."""
+
+    def __init__(
+        self, in_channels: int, out_channels: int, msg_dim: int, time_enc: TimeEncoder
+    ):
+        """Initialize temporal attention layer."""
+        super().__init__()
+        self.time_enc = time_enc
+        edge_dim = msg_dim + time_enc.out_channels
+        self.conv = TransformerConv(
+            in_channels, out_channels // 2, heads=2, dropout=0.1, edge_dim=edge_dim
+        )
+
+    def forward(self, x, last_update, edge_index, t, msg):
+        """Compute node embeddings."""
+        rel_t = last_update[edge_index[0]] - t
+        rel_t_enc = self.time_enc(rel_t.to(x.dtype))
+        edge_attr = torch.cat([rel_t_enc, msg], dim=-1)
+        return self.conv(x, edge_index, edge_attr)
+
+
+class LinkPredictor(torch.nn.Module):
+    """Compute probabilities for edges between sources and destinations."""
+
+    def __init__(self, in_channels):
+        """Initialize link predictor."""
+        super().__init__()
+        self.lin_src = Linear(in_channels, in_channels)
+        self.lin_dst = Linear(in_channels, in_channels)
+        self.lin_final = Linear(in_channels, 1)
+
+    def forward(self, z_src, z_dst):
+        """Compute edge probabilities."""
+        h = self.lin_src(z_src) + self.lin_dst(z_dst)
+        h = h.relu()
+        return self.lin_final(h)
+
+
+graph = MOOC()
+
+
+def train_val_test_batches(test_ratio: float = 0.1, val_ratio: float = 0.1):
+    """Split the graph edges into train/val/test batches."""
+    edges = np.loadtxt(
+        os.path.join(
+            graph.data_dir(), "raw", graph.GRAPH_NAME, f"{graph.GRAPH_NAME}.csv"
+        ),
+        delimiter=",",
+        skiprows=1,
+        usecols=(0, 1, 2),
+    ).astype(np.int64)
+
+    edges[:, 1] += graph.max_src
+    val_time, test_time = np.quantile(
+        edges[:, 2], [1 - test_ratio - val_ratio, 1 - test_ratio]
+    )
+
+    test_idx = (edges[:, 2] <= test_time).sum()
+    val_idx = (edges[:, 2] <= val_time).sum()
+    return edges[:val_idx], edges[val_idx:test_idx], edges[test_idx:]
+
+
+train_data, _, test_data = train_val_test_batches(0.15, 0.15)
+train_loader = DataLoader(train_data, batch_size=train_batch_size)
+test_loader = DataLoader(test_data, batch_size=train_batch_size)
+
+memory_dim = time_dim = embedding_dim = 100
+
+memory = TGNMemory(
+    graph.num_nodes,
+    graph.edge_feature_dim,
+    memory_dim,
+    time_dim,
+    message_module=IdentityMessage(graph.edge_feature_dim, memory_dim, time_dim),
+    aggregator_module=LastAggregator(),
+).to(device)
+
+gnn = GraphAttentionEmbedding(
+    in_channels=memory_dim,
+    out_channels=embedding_dim,
+    msg_dim=graph.edge_feature_dim,
```

Comment:
This code belongs in an if __name__ == '__main__', not in file

Reply:
There are way too many moving parts here and putting everything under __main__ makes code way longer.

Reply:
How so? You can put train/test functions in main if they use the globals

Reply:
you don't need check module name then if you want to keep globals.